Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt

```diff
@@ -0,0 +1,249 @@
+import de.rki.coronawarnapp.server.protocols.internal.AttenuationDurationOuterClass.AttenuationDuration
+import de.rki.coronawarnapp.storage.LocalData
+import de.rki.coronawarnapp.storage.RiskLevelRepository
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.millisecondsToHours
+import kotlinx.coroutines.flow.first
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.math.round
+
+@Singleton
+class DefaultRiskLevels @Inject constructor(
+    private val appConfigProvider: AppConfigProvider,
+    private val enfClient: ENFClient
+) : RiskLevels {
+
+    override fun updateRepository(riskLevel: RiskLevel, time: Long) {
+        val rollbackItems = mutableListOf<RollbackItem>()
+        try {
+            Timber.tag(TAG).v("Update the risk level with $riskLevel")
+            val lastCalculatedRiskLevelScoreForRollback =
+                RiskLevelRepository.getLastCalculatedScore()
+            updateRiskLevelScore(riskLevel)
+            rollbackItems.add {
+                updateRiskLevelScore(lastCalculatedRiskLevelScoreForRollback)
+            }
+
+            // risk level calculation date update
+            val lastCalculatedRiskLevelDate = LocalData.lastTimeRiskLevelCalculation()
+            LocalData.lastTimeRiskLevelCalculation(time)
+            rollbackItems.add {
+                LocalData.lastTimeRiskLevelCalculation(lastCalculatedRiskLevelDate)
+            }
+        } catch (error: Exception) {
+            Timber.tag(TAG).e(error, "Updating the RiskLevelRepository failed.")
+
+            try {
+                Timber.tag(TAG).d("Initiate Rollback")
+                for (rollbackItem: RollbackItem in rollbackItems) rollbackItem.invoke()
+            } catch (rollbackException: Exception) {
+                Timber.tag(TAG).e(rollbackException, "RiskLevelRepository rollback failed.")
+            }
+
+            throw error
+        }
+    }
+
+    override fun calculationNotPossibleBecauseOfOutdatedResults(): Boolean {
+        // if the last calculation is longer in the past as the defined threshold we return the stale state
+        val timeSinceLastDiagnosisKeyFetchFromServer =
+            TimeVariables.getTimeSinceLastDiagnosisKeyFetchFromServer()
+                ?: throw RiskLevelCalculationException(
+                    IllegalArgumentException(
+                        "Time since last exposure calculation is null"
+                    )
+                )
+        /** we only return outdated risk level if the threshold is reached AND the active tracing time is above the
+        defined threshold because [UNKNOWN_RISK_INITIAL] overrules [UNKNOWN_RISK_OUTDATED_RESULTS] */
+        return timeSinceLastDiagnosisKeyFetchFromServer.millisecondsToHours() >
+            TimeVariables.getMaxStaleExposureRiskRange() && isActiveTracingTimeAboveThreshold()
+    }
+
+    override fun calculationNotPossibleBecauseNoKeys() =
```

Comment:
I would rename this function to calculationNotPossibleBecauseOfNoKeys, to keep the naming consistent with the function above.